engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/HashMapInsertNode.java

```diff
@@ -29,9 +30,28 @@ public static HashMapInsertNode build() {
   public abstract EnsoHashMap execute(Object self, Object key, Object value);
 
   @Specialization
+  @TruffleBoundary
   EnsoHashMap doEnsoHashMap(EnsoHashMap hashMap, Object key, Object value) {
-    hashMap.getMapBuilder().add(key, value);
-    return hashMap.getMapBuilder().build();
+    EnsoHashMapBuilder mapBuilder = hashMap.getMapBuilder();
+    boolean containsKey = mapBuilder.get(key) != null;
+    boolean insertCalledOnMap = hashMap.isInsertCalled();
+    if (insertCalledOnMap || containsKey) {
+      // insert was already called on this map => We need to duplicate MapBuilder
+      // If a key is already contained in the Map there is no way telling whether there is another
+      // binding pointing to the Map, and we do not want to mutate this older binding.
+      var newMapBuilder = mapBuilder.duplicate();
+      if (hashMap.getHashSize() < mapBuilder.getSize()) {
+        newMapBuilder.removeLastEntries(mapBuilder.getSize() - hashMap.getHashSize());
+      }
```

Comment:
Assume that `mapBuilder.getSize == 10k` and `hashMap.getHashSize() == 100`. On `mapBuilder.duplicate` we'd copy the 10k elements just to remove the 9900 of them in the next call.

Maybe such scenario will not be very prevalent, but could we not add `duplicateFirstNEntries` to `mapBuilder` and already only copy what we need (the 100 entries in the example above)?

OK to do that in a later pass too, I really don't want to postpone merge unnecessarily. But it feels like a lot of unnecessary work for the CPU.

Reply:
Good idea. Introduced `EnsoHashMapBuilder.duplicatePartial` method, so that we don't have to copy everything and then remove it. I think it even makes the code a bit more readable.